tools/fm-editor/graph/FeatureModelGraph.cpp

```diff
@@ -0,0 +1,198 @@
+  scaleView(pow(2., -Event->angleDelta().y() / 240.0));
+}
+#endif
+
+void FeatureModelGraph::drawBackground(QPainter *Painter, const QRectF &Rect) {
+  Q_UNUSED(Rect);
+
+  // Shadow
+  QRectF SceneRect = this->sceneRect();
+  QRectF RightShadow(SceneRect.right(), SceneRect.top() + 5, 5,
+                     SceneRect.height());
+  QRectF BottomShadow(SceneRect.left() + 5, SceneRect.bottom(),
+                      SceneRect.width(), 5);
+  if (RightShadow.intersects(Rect) || RightShadow.contains(Rect)) {
+    Painter->fillRect(RightShadow, Qt::darkGray);
+  }
+  if (BottomShadow.intersects(Rect) || BottomShadow.contains(Rect)) {
+    Painter->fillRect(BottomShadow, Qt::darkGray);
+  }
+
+  // Fill
+  QLinearGradient Gradient(SceneRect.topLeft(), SceneRect.bottomRight());
+  Gradient.setColorAt(0, Qt::white);
+  Gradient.setColorAt(1, Qt::lightGray);
+  Painter->fillRect(Rect.intersected(SceneRect), Gradient);
+  Painter->setBrush(Qt::NoBrush);
+  Painter->drawRect(SceneRect);
+
+  QFont Font = Painter->font();
+  Font.setBold(true);
+  Font.setPointSize(14);
+  Painter->setFont(Font);
+  Painter->setPen(Qt::lightGray);
+  Painter->setPen(Qt::black);
+}
+void FeatureModelGraph::scaleView(qreal ScaleFactor) {
+  qreal Factor = transform()
+                     .scale(ScaleFactor, ScaleFactor)
+                     .mapRect(QRectF(0, 0, 1, 1))
+                     .width();
+  if (Factor < 0.07 || Factor > 100) {
+    return;
+  }
+
+  scale(ScaleFactor, ScaleFactor);
+}
+void FeatureModelGraph::zoomIn() { scaleView(qreal(1.2)); }
+void FeatureModelGraph::zoomOut() { scaleView(1 / qreal(1.2)); }
+
+
+FeatureNode* FeatureModelGraph::addNode(Feature* Feature, FeatureNode* Parent) {
+  auto NewNode = std::make_unique<FeatureNode>(Feature);
+  auto * NewEdge = new FeatureEdge(Parent,NewNode.get());
+  scene()->addItem(NewEdge);
+  scene()->addItem(NewNode.get());
+  auto NewNodeRaw = NewNode.get();
+  Nodes.push_back(std::move(NewNode));
+  auto NextChildren =std::vector<FeatureNode*>(EntryNode->children().size());
+  auto CurrentChildren = EntryNode->children();
+  std::transform(CurrentChildren.begin(),CurrentChildren.end(),NextChildren.begin(),[](FeatureEdge* Edge){return Edge->targetNode();});
+  positionRec(1,NextChildren,5);
+  EntryNode->setPos(EntryNode->childrenWidth()/2,10);
+  return NewNodeRaw;
+}
+
+
+FeatureNode* FeatureModelGraph::getNode(std::string Name) {
+  auto It = std::find_if(Nodes.begin(),Nodes.end(),[&Name](auto const &Node){return Node->getName() == Name;});
+  if (It != Nodes.end()) {
+    return It->get();
+  }
+  return nullptr;
+
+}
+void FeatureModelGraph::deleteNode(bool Recursive, FeatureNode* Node) {
+  auto *Parent = Node->parent()->sourceNode();
+  if(!Recursive){
+    for(auto *Child: Node->children()){
+    Child->setSourceNode(Parent);
+    }
+
+    Node->children().clear();
+  }else {
+    for(auto *Child: Node->children()){
+    deleteNode(true,Child->targetNode());
+    }
+  }
```

Comment:
```suggestion
  if(Recursive) {
    for(auto *Child : Node->children()){
      deleteNode(true,Child->targetNode());
    }
  } else {
    for(auto *Child: Node->children()) {
      Child->setSourceNode(Parent);
    }

    Node->children().clear();
  }
```
+ extra formatting I cannot do by hand